allennlp/models/encoder_decoders/nlvr_semantic_parser.py

```diff
@@ -0,0 +1,365 @@
+                 sentence_embedder: TextFieldEmbedder,
+                 nonterminal_embedder: TextFieldEmbedder,
+                 terminal_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 decoder_trainer: DecoderTrainer,
+                 decoder_beam_search: BeamSearch,
+                 max_decoding_steps: int,
+                 attention_function: SimilarityFunction) -> None:
+        # Forcing embed_terminals to be True because we are not doing entity linking for NLVR
+        # (atleast for now).
+        super(NlvrSemanticParser, self).__init__(vocab=vocab,
+                                                 question_embedder=sentence_embedder,
+                                                 nonterminal_embedder=nonterminal_embedder,
+                                                 terminal_embedder=terminal_embedder,
+                                                 encoder=encoder,
+                                                 decoder_trainer=decoder_trainer,
+                                                 decoder_beam_search=decoder_beam_search,
+                                                 max_decoding_steps=max_decoding_steps,
+                                                 attention_function=attention_function,
+                                                 embed_terminals=True)
+
+        nonterminal_embed_dim = nonterminal_embedder.get_output_dim()
+        action_embedding_dim = nonterminal_embed_dim * 2
+
+        # self._decoder_step would be set to ``WikiTablesDecoderStep``. Rewriting it.
+        self._decoder_step = NlvrDecoderStep(encoder_output_dim=self._encoder.get_output_dim(),
+                                             action_embedding_dim=action_embedding_dim,
+                                             attention_function=attention_function)
+    @overrides
+    def forward(self,
+                sentence: Dict[str, torch.LongTensor],
+                world: List[NlvrWorld],
+                actions: List[List[ProductionRuleArray]],
+                agenda: torch.LongTensor,
+                label: torch.LongTensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ,unused-argument
+        """
+        Decoder logic for producing type constrained target sequences, that maximize coverage of
+        their respective agendas. This will change soon, to include a denotation based score as
+        well, once we have a way to transform action sequences into logical forms that can be
+        executed to produce denotations.
+        """
+        # TODO(pradeep): Use labels.
+        embedded_input = self._question_embedder(sentence)
+        sentence_mask = nn_util.get_text_field_mask(sentence).float()
+
+        batch_size = embedded_input.size(0)
+
+        encoder_outputs = self._encoder(embedded_input, sentence_mask)
+
+        final_encoder_output = encoder_outputs[:, -1]  # (batch_size, encoder_output_dim)
```

Comment:
Assuming the encoder is a recurrent network, wouldn't it just pass along the state if the input at a timestep is masked?

Reply:
I don't think you can do this - it is masked with zeros, so you have to `index_select` the lengths of the sequences. see here: https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2vec_encoders/pytorch_seq2vec_wrapper.py#L100

Reply:
Ah! I see, thanks! That is not how I imagined the encoder was implemented. Why is the padding zeros and not a repetition of the final encoder state?

Reply:
:man_shrugging: That's how it's implemented in pytorch, I don't know. It's not obvious to me that having non-zero padded states is actually better.

Reply:
Repeating the last state makes more sense to me because of what I said in my earlier comment about what I would expect from a recurrent network when the input is masked. Keras does it that way, and my thinking may simply be biased towards it.
Anyway, thanks for the pointer!